### ISSUE TITLE
Allow 'root' opt passed in data

### ISSUE THREAD

**Issue by lext-7** [mde/ejs#224]
```javascript
var _OPTS = ['delimiter', 'scope', 'context', 'debug', 'compileDebug',
  'client', '_with', 'rmWhitespace', 'strict', 'filename',
  'root'
];
```

**Comment by RyanZim** [comment]
@lext-7 We don't want to do this, since this introduces a security issue when the data is untrusted. This could potentially allow an attacker to access private files on the server. Is there a specific reason why you can't pass your data & options separately?

**Comment by nwoltman** [comment]
I'm not sure about @lext-7's use case, but I'd like to be able to specify the `root` option when using Express. Perhaps it could be allowed the same [way as the `cache` option](https://github.com/mde/ejs/blob/master/lib/ejs.js#L347).

**Comment by RyanZim** [comment]
@nwoltman I see what you're saying; however, that won't work, since an express user may be getting user-generated content for the data. Making an exception for express would open a security hole.

@mde Thoughts on this?

**Comment by mde** [comment]
@nwoltman Does Express give any sort of interface/allowance for configuring the templating engine?

**Comment by nwoltman** [comment]
@mde Not really. Express only ever calls `ejs.renderFile()` directly. The closest thing there is to being able to set some sort of configuration is using [`app.set()`](http://expressjs.com/en/4x/api.html#app.set) to set the `'view options'` setting which EJS reads [here](https://github.com/mde/ejs/blob/master/lib/ejs.js#L354) (even though there's no documentation for that setting anywhere).

**Comment by nwoltman** [comment]
Actually, I can pass in options to EJS by using a custom render function with Express:

```js
const app = express();
const ejsOptions = {
  cache: app.get('view cache'),
  root: '/path/to/views',
};

app.engine('ejs', (path, data, cb) => {
  ejs.renderFile(path, data, ejsOptions, cb);
});
```

This solves my problem and is probably the best way to configure EJS with Express rather than setting the completely undocumented `'view options'` setting on the `app`.

**Comment by mde** [comment]
Thank you so much for unraveling this! We need to add this to docs, so other people know how to do bespoke config for EJS using Express.

**Comment by tkesgar** [comment]
Just adding some more information for everyone wondering how to set EJS options in Express.

It seems that the "correct" way (that is, as done by most other template engines) to set template engine options is to set it in `app.locals`:

```js
app.locals.delimiter = '?'
app.locals.strict = false
```

At `__express(filename, options, callback)`, `options` will contain the values from `res.locals`, `app.locals`, and also some other variables such as `cache` (I can't find any documentation on the other variables).

`view options` mentioned above is documented in Express 2, but seems to be not used anymore since 3.

Maybe a little thing to consider is `compileDebug` flag, which Pug determines based on `NODE_ENV`: https://github.com/pugjs/pug/blob/703383122621828cff74950d2e6abbe8deb87cc5/packages/pug/lib/index.js

**Comment by nwoltman** [comment]
@tkesgar Just to be clear, that won't work for the `root` option.

Explantion:

Options set in `app.locals` are merged into the `data` argument that is passed to `renderFile()`, and only "trusted" options are allowed to be passed in the `data` object. Options that are not "trusted" (such as `root`) must be passed in the `opts` argument, which isn't possible when using the `__express()` function because:

```js
__express(filename, options, callback)
// maps to
renderFile(path, data, cb)
// but we need
renderFile(path, data, opts, cb)
```

See the [`renderFile()` definition](https://github.com/mde/ejs/blob/master/lib/ejs.js#L367-L380).

**Comment by nwoltman** [comment]
> `view options` mentioned above is documented in Express 2, but seems to be not used anymore since 3.

That means that [these comments]() are totally wrong and EJS should consider deprecating usage of `view options`.

**Comment by mde** [comment]
I have pushed a fix for all this stuff to NPM in v2.5.9. Should make docs much less necessary, although there's probably still some use for good docs, given that there are at least four different ways you can do this now. Also, unsafe options like `root` cannot be set by being passed in the actual function call -- probably worth letting people know.
